authoring-recipes/types-of-recipes.md

```diff
@@ -0,0 +1,88 @@
+
+Refaster templates can also be used as a starting point for more complex recipe implementations. This is due to the fact that, when you define a refaster template, an imperative recipe is what is actually created behind the scenes. You can find these imperative recipes in the `build` directory after you've built your repository.
+
+A variety of refaster templates can be found in the [StringRules](https://github.com/openrewrite/rewrite-migrate-java/blob/v2.1.1/src/main/java/org/openrewrite/java/migrate/lang/StringRules.java#L23-L48) class in `rewrite-migrate-java`:
+
+```java
+@RecipeDescriptor(
+        name = "Replace redundant `String` method calls with self",
+        description = "Replace redundant `substring(..)` and `toString()` method calls with the `String` self."
+)
+@SuppressWarnings("StringOperationCanBeSimplified")
+public static class RedundantCall {
+    @BeforeTemplate
+    public String start(String string) {
+        return string.substring(0, string.length());
+    }
+
+    @BeforeTemplate
+    public String startAndEnd(String string) {
+        return string.substring(0);
+    }
+
+    @BeforeTemplate
+    public String toString(String string) {
+        return string.toString();
+    }
+
+    @AfterTemplate
+    public String self(String string) {
+        return string;
+    }
+}
+```
+
+## Imperative recipes
+
+[Imperative recipes](https://docs.openrewrite.org/authoring-recipes/writing-a-java-refactoring-recipe) offer the most freedom and functionality at the cost of being more difficult to create. They allow you to write Java code to implement your recipe.
+
+Imperative recipes are ideal for situations where there is a lot of complexity or nuance needed to determine what should be changed or what it should be changed into. For instance, if you wanted to write a recipe that added the `final` modifier to any local variables that aren't reassigned, you would need to create an imperative recipe as refaster templates don't have the ability to determine whether a variable has been reassigned.
+
+If your recipe can be defined via a declarative YAML file or via a refaster template, you should do so -- even if you can technically create an imperative recipe that does the same thing.
+
+The [FinalizeLocalVariables recipe](https://github.com/openrewrite/rewrite-static-analysis/blob/main/src/main/java/org/openrewrite/staticanalysis/FinalizeLocalVariables.java) is a great example of an imperative recipe.
+
+## Next steps
+
+* If you want more information about the types of recipes and the recipe execution pipeline, please see our [recipes documentation](/concepts-and-explanations/recipes.md)
+* If you want to write your own imperative recipe, please check out the [writing a Java refactoring recipe](/authoring-recipes/writing-a-java-refactoring-recipe.md) guide.
```

Comment:
MIght be good to link to how to test recipes here as well, as that is a likely first step folks might want to take. Perhaps we should even spin that out into a separate doc, as it can be shared between declarative, refaster and imperative recipes.

Reply:
I added a link to the existing testing guide. We can potentially update that in the future to include more information on testing.